### DISCUSSION TITLE
How much strain can sqlpage.exe handle?

### DISCUSSION THREAD

**Discussion by Pieter3033** [sqlpage/SQLPage#344] [Q&A] [answered]
Hi @lovasoa and SQLpage community.

I was wondering how much strain/ requests can sqlpage.exe handle.

I have written multiple internal and external Web Applications for the company I am working for.
Currently there is over 100 active users daily across the multiple Web Applications, I am concerned about the processing of these requests through SQLpage.exe because all of these Web Applications are running on 1 sqlpage.exe terminal.

Just to give insight of the Web Applications that are running on a single terminal.

1. Technician. (At most 44 users daily)
2. SRF. (10 users daily)
3. Unit inspection program. (Approximately 30 users daily)
4. Client App 1 (2-6 users daily)
5. Client App 2 (5-10 users daily)
6. Client App 3 (Unknown count of users)

Unfortunately I am unable to give a amount of request that is being processed daily, because there are too much.

I have experienced this Errors before, but I am not sure if it is related to my above query.

`[2024-05-23T14:30:47.690Z ERROR sqlpage::webserver::http] Stopping rendering early because we were unable to flush data to client: The HTTP response writer with a capacity of 128 has already been closed: channel closed`

` [2024-05-23T14:30:47.691Z ERROR sqlpage::webserver::http] Could not flush data to client: channel closed 2024-05-23T14:30:47.695Z ERROR sqlpage::webserver::http] Stopping rendering early because we were unable to flush data to client: The HTTP response writer with a capacity of 128 has already been closed: channel closed [2024-05-23T14:30:47.696Z ERROR sqlpage::webserver::http] Could not flush data to client: channel closed`

**Comment by lovasoa** [comment, marked as answer]
Hi @Pieter3033 ! 

It's very nice to hear about how your applications have grown ! I still remember our first call when you were just getting started and hadn't deployed your app. I'd love to get in touch again and to learn about your new apps if you are down for it: `contact at ophir dot dev`

SQLPage should be able to easily handle at least tens of requests per second in a production setting, and hundreds of requests per second on smaller pages, depending on your hardware.  

The error messages seem to be linked to users closing the connection before the page has fully loaded. This is nothing to worry about. Have users reported display issues in the applications ?

**Reply by Pieter3033** [reply]
Thanks for the answer @lovasoa.

I'd love to get in touch again, I believe you'd be surprised how much it impacted our company and how much I have achieved with no prior web development knowledge.

I'd also like to ask you a few questions regarding SQLpage, but I'll contact you and we can arrange a call sometime.

**Reply by lovasoa** [reply]
Hi @Pieter3033 ! Do you think we can have a call during the upcoming week?

**Reply by Pieter3033** [reply]
Hi @lovasoa , yes this week will work fine.
What time will fit you best?

**Reply by lovasoa** [reply]
I sent you an invite for tomorrow morning

**Comment by Pieter3033** [comment]
For interest sake.

**Network usage**
Max
![image](https://github.com/user-attachments/assets/f03bf386-6561-435f-8e21-e00d4ab8f9d7)

Min:
![image](https://github.com/user-attachments/assets/7b5e75b6-a84d-4d8e-bf9f-f883b8082a9b)

**Memory Usage**
![image](https://github.com/user-attachments/assets/ec2ef18e-064a-4db7-a803-c40779a29378)

**CPU Usage**
![image](https://github.com/user-attachments/assets/c27b1796-11be-4177-93f4-7dfe43dd0271)


Operating System: Windows server

Very impressed with how smooth SQLPage is running, with the amount of traffic.

**Reply by Pieter3033** [reply]
Ooh and 
**Disk usage**
![image](https://github.com/user-attachments/assets/0bee14f8-012c-4d81-984d-e65fc9b17aaa)

**Reply by lovasoa** [reply]
I love this ! Thanks for sharing :muscle:

**Reply by lovasoa** [reply]
Do you have an idea how many concurrent users you have ? How many requests per seconds during peak load ?

**Reply by Pieter3033** [reply]
> Do you have an idea how many concurrent users you have ? How many requests per seconds during peak load ?

I'll have a look on Monday around 12:00 and Wednesday around 12:00 as well.

Is there a way to pull a logfile on sqlpage.exe, or maybe an alternative way to count the amount of request lets say per minute?

**Reply by lovasoa** [reply]
Yes, when you launch sqlpage from the shell, you can [redirect its output to a file](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_redirection?view=powershell-7.4).

But you probably want to run it as a windows service (to handle start on boot, restarts, ...), in which case you can use [nssm](https://nssm.cc/usage) to handle the service and redirect he logs to a file, that you can then check for errors, and use to count requests.

![image](https://github.com/user-attachments/assets/fe948ddb-cf41-464a-93e7-222468eed3e0)
